In [4]:
import numpy as np
def sigmoid(x) :
    return 1 / (1 + np.exp(-x))

def softmax(a) :
    c = np.max(a)
    exp_a = np.exp(a-c)
    sum_exp_a = np.sum(exp_a)
    y = exp_a / sum_exp_a

    return y

In [5]:
import sys, os
sys.path.append(os.pardir)
from dataset.mnist import load_mnist

(x_train, t_train),(x_test,t_test) = \
    load_mnist(flatten=True,normalize=False)

# 각 데이터의 형상 출력
print(x_train.shape)
print(t_train.shape)
print(x_test.shape)
print(t_test.shape)



(60000, 784)
(60000,)
(10000, 784)
(10000,)


In [6]:
import sys,os
sys.path.append(os.pardir)
import numpy as np
from dataset.mnist import load_mnist
from PIL import Image
import pickle

def img_show(img):
    pil_img = Image.fromarray(np.uint8(img))
    pil_img.show()

(x_train, t_train),(x_test,t_test) = \
    load_mnist(flatten=True,normalize=False)

img = x_train[2]
label = t_train[2]
print(label)

print(img.shape)
img = img.reshape(28,28)
print(img.shape)

img_show(img)

4
(784,)
(28, 28)


## 신경망의 추론 처리
이 신경망은 입력층 뉴런을 784개, 출력층 뉴런을 10개로 구성한다.
입력층 뉴런이 784개인 이유는 이미지 크기가 28x28 이기 때문이고, 출력층 뉴런이 10개인 이유는 이 문제가 0에서 9까지의 숫자를 구분하는 문제이기 때문이다.
한편 은닉층은 총 두 두개로, 첫 번째 은닉층에서는 50개의 뉴런을, 두 번째 은닉층에는 100개의 뉴런을 배치할 것이다.
여기서 50과 100은 임의로 정한 값이다.

In [11]:
def get_data():
    (x_train,t_train),(x_test,t_test) = \
        load_mnist(normalize=True, flatten=True,one_hot_label=False)
    return x_test,t_test

def init_network():
    with open("sample_weight.pkl",'rb') as f:
        network = pickle.load(f)

    return network

def predict(network, x):
    W1,W2,W3 = network['W1'],network['W2'],network['W3']
    b1,b2,b3 = network['b1'],network['b2'],network['b3']

    a1 = np.dot(x,W1) + b1
    z1 = sigmoid(a1)
    a2 = np.dot(z1,W2) + b2
    z2 = sigmoid(a2)
    a3 = np.dot(z2,W3) + b3
    y = softmax(a3)

    return y


In [17]:
x,t = get_data()
network = init_network()

accuracy_cnt = 0
print(np.shape(x))
for i in range(len(x)):
    y = predict(network,x[i])
    p = np.argmax(y)
    if p == t[i]:
        accuracy_cnt += 1

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))

(10000, 784)
(10000,)
Accuracy:0.9352


In [13]:
x,y = get_data()
network = init_network()

batch_size = 100 #배치 크기
accuracy_cnt = 0

for i in range(0,len(x),batch_size):
    x_batch = x[i:i+batch_size]
    y_batch = predict(network,x_batch)

    p = np.argmax(y_batch,axis=1)
    accuracy_cnt += np.sum(p == t[i:i+batch_size])

print("Accuracy:" + str(float(accuracy_cnt) / len(x)))

Accuracy:0.9352
